### ISSUE TITLE
How do you return posts created in a date range?

### ISSUE THREAD

**Issue by davydnorris** [nuxt/content#437]
Hi all,

I am trying to implement an archives search in my blog, which requires me to return posts within a date range. I am trying the following but with no luck:

```
        let postQuery = this.$content('blog', { deep: true })
          .only(['dir', 'title', 'author', 'description', 'image', 'createdAt'])
          .sortBy('createdAt', 'desc');

        if (this.dateFilter) {
          const startDate = new Date(this.dateFilter.setDate(1));
          const endDate = new Date(this.dateFilter.setMonth(this.dateFilter.getMonth() + 1, 1));
          postQuery = postQuery.where({ createdAt: { $between: [startDate, endDate] } });
        }

        this.posts = await postQuery.fetch();
```
the above gives no errors, but just doesn't return anything. I have tried all sorts of variations on this, such as:
```
          postQuery = postQuery.where({
            $and: [
              { createdAt: { $gte: startDate } },
              { createdAt: { $lte: endDate } }
            ]
          });
```
but still no luck. The only thing I can't think of is that createdAt is not a date?

Any help much appreciated!

**Comment by azrikahar** [comment]
This is not the final solution, but I'm facing some weird interaction regarding the date here.  It works upon refresh on `blog` page, but using button to trigger fetch will fail instead. CodeSandbox link: https://codesandbox.io/s/happy-jepsen-1x3mo?file=/pages/blog/index.vue

## Scenario

blog posts:
- Post 1 is created at 2020-08-25
- Post 2 is created at 2020-08-31

I'm using the Nuxt fetch() method instead of asyncData() here so we can trigger $fetch on button click. If we were to go to `/blog` via links or routes, we will see the fetch returning empty, meaning the filtering is incorrect. The fetch didn't fail because it would show `Unable to fetch blog posts` if it did.

However, if you go to `/blog`, then do a refresh by clicking on the refresh button on the preview box (not our browser, the sandbox preview), you can see the fetch worked and it returned only Post 2 because I hardcoded the initial start date as `2020-08-26` which is greater than the Post 1's `2020-08-25`.

I've also noticed if the fetch happened on the server, it just needs `new Date(date)` to console.log an ISO date format, whereas the client side fetch will need `new Date(date).toISOString()` to have the same format. I thought having the same format for client side will fix it, but it still fails :/ Not really sure what is happening here. 

I might be doing something wrong, but hoping that this can give some clue. You can also try changing the `startDate` on the data option to `2020-08-25`, then do a refresh to see that it works and shows both posts for you.

**Comment by davydnorris** [comment]
Hmmm... that looks as thought the filter is doing a text compare instead of a date compare.

Maybe I should turn this into an Issue?

**Comment by davydnorris** [comment]
I've done some more digging and it appears that the createdAt property is getting turned into a String somewhere along the line, which may be why the Date compare routines are not working properly. This is the result of the fetch:

```
{
    title: 'Blog #4 Headline',
    description: 'Lorem Ipsum type stuff',
    createdAt: '2020-02-17T00:00:00.000Z',
    image: 'IMG_6263.JPG',
    dir: '/blog/2020-02-17-blog-number-4',
    path: '/blog/2020-02-17-blog-number-4/index',
    extension: '.md'
  },
```

**Comment by davydnorris** [comment]
I just tried a very basic custom filter as per LokiJS documentation and it appears it's being ignored completely :-( - that would have worked beautifully if it ran.

```
postQuery = postQuery.where((obj) => { console.log(obj); return true; });
```

**Comment by azrikahar** [comment]
> I've done some more digging and it appears that the createdAt property is getting turned into a String somewhere along the line,

I was also wondering about this. I tried console.logging the value of the start date that is used for the query,  They are both the typeof object, but the "on page refresh" one logs `2020-08-26T00:00:00.000Z` (this works) whereas the client side triggered one will log `Wed Aug 26 2020 08:00:00 GMT+0800 (Malaysia Time)` instead which does not work.

> I just tried a very basic custom filter as per LokiJS documentation and it appears it's being ignored completely :-( - that would have worked beautifully if it ran.

I did try the other comparator like `$jgte` instead of `$gte` to no avail.

---

With that said, since the current implementation uses `stats.birthTime` which is a Date object, perhaps an alternative is allow us to configure it as `stats.birthTimeMs` which is the epoch time? That way we can compare the numeric epoch timestamp without any issue. I recall seeing this suggestion probably in the LokiJS issues, but can't seem to find it back again.

**Comment by azrikahar** [comment]
I cracked the code! (and probably my head as well)

### CodeSandBox Link: https://codesandbox.io/s/solitary-tree-sf9ew?file=/pages/blog/index.vue

As I mentioned the typeof thing and the using Epoch in the library, it dawned on me on why not trying to use epoch on our frontend code? Seems like it actually works! We just need to use `new Date(date).valueOf()` instead of the typical `new Date(date)` and we will have a working filter. Setting both of their value to empty string also seems to work without needing to cater cases like one of it is filled but the other not filled and might interfere with the $and query. But no, if one of them is empty, it will act as if it is not bound and we can just set startDate or endDate only. Hope it's not just a "work in my machine/env" type of deal here.

**Comment by davydnorris** [comment]
Well done!!

This works for me too, as does 

`postQuery = postQuery.where({ createdAt: { $between: [startDate.valueOf(), endDate.valueOf()] } });`

**Comment by azrikahar** [comment]
Yea I was using $between as well, but then I noticed setting 1st Sept as end date didnt show my Post 2, so I thought it is $between not being inclusive. However I then realized its the endDate being 00:00 of 1st Sept, thats why it wasn't showing Post 2, not the $between's fault.

Glad we can resolve this :) perhaps it can be documented in the snippets section?

**Comment by gsjen** [comment]
`valueOf()` only worked for me in development. In production I had to switch to `toJSON()`.

**Comment by davydnorris** [comment]
Weird - mine's working fine in production

**Comment by szwacz** [comment]
I don't fully get it, so this is a bug or desired behaviour?

I see Loki has the option to put simple function into `where`:
```js
var results = coll.where(function(obj) {
  return obj.legs === 8;
});
```
I've tried and looks like this is not suppoerted by nuxt-content. Wouldn't this be a neat solution for this problem? Then you can manipulate your dates however you want.

**Comment by ManasMadrecha** [comment]
@szwacz 
I think Content uses `find` features of Loki, rather than `where`, though in Content it is called `where`. 🤔

**Comment by adam-buckley** [comment]
> I cracked the code! (and probably my head as well)
> 
> ### CodeSandBox Link: https://codesandbox.io/s/solitary-tree-sf9ew?file=/pages/blog/index.vue
> As I mentioned the typeof thing and the using Epoch in the library...

Thank you so much for figuring this out! Been trying for ages trying to get this to work!

**Comment by oneminch** [comment]
It appears a fix was merged for this and it has been documented for the older version of Nuxt Content (#2704). If you've since migrated to Nuxt Content v2/v3, the querying API has evolved a little and now comes with a SQL style `BETWEEN` operator.

Can this issue be closed @davydnorris?

**Comment by davydnorris** [comment]
> It appears a fix was merged for this and it has been documented for the older version of Nuxt Content ([#2704](https://github.com/nuxt/content/pull/2704)). If you've since migrated to Nuxt Content v2/v3, the querying API has evolved a little and now comes with a SQL style `BETWEEN` operator.
> 
> Can this issue be closed [@davydnorris](https://github.com/davydnorris)?

Yes - I didn't realise it was still open. We found a workaround some years ago, but this is a great update
